File: testdata/precision-suite-kotlin/tp_cmdinjection.kt
// Command injection: a user-controlled request parameter is concatenated into a
// shell command string and handed to Runtime.getRuntime().exec — no argument
// vector, no allow-list. This is CWE-78. A correct scanner fires TAINT-002.
// nox's Kotlin taint model sees `name` tainted by request.getParameter reach the
// .exec sink (matched on the method suffix) with no sanitizer on the path.
package com.example

import javax.servlet.http.HttpServletRequest

class ReportRunner {
    fun generate(request: HttpServletRequest) {
        val name = request.getParameter("report")
        val cmd = "generate-report --name " + name
        Runtime.getRuntime().exec(cmd) // nox-expect: TAINT-002
    }
}

File: testdata/precision-suite-scala/tp_cmdinjection.scala
// Command injection: a Play request query parameter is interpolated into a shell
// command run via scala.sys.process — CWE-78. A correct scanner fires TAINT-002.
// The tainted value is spliced into the command string rather than passed as a
// discrete argument vector, which is the injectable form.
package controllers

import scala.sys.process._

class ReportController {
  // generate runs an external tool named partly by untrusted input — the bug.
  def generate(request: Request): Int = {
    val name = request.getQueryString("report")
    val cmd = "generate-report " + name
    Process(cmd).run().exitValue() // nox-expect: TAINT-002
  }

  // bang runs the command via the postfix `.!` process operator — the same sink
  // reached through the scala.sys.process DSL rather than a Process(...) call.
  def bang(request: Request): Int = {
    val host = request.getQueryString("host")
    s"ping -c 1 $host".! // nox-expect: TAINT-002
  }
}
